pkgs/test/nixpkgs-check-by-name/src/eval.rs

```diff
@@ -96,10 +97,19 @@ pub fn check_values(
         command.arg("-I");
         command.arg(path);
     }
-
-    let result = command
-        .output()
-        .context(format!("Failed to run command {command:?}"))?;
+    command.arg("-");
+    let mut child = command
+        .spawn()
+        .unwrap_or_else(|err| panic!("failed to run command '{:?}': {}", stringify!(command), err));
+    let mut stdin = child.stdin.take().expect("Failed to open stdin");
+    std::thread::spawn(move || {
+        stdin
+            .write_all(EXPR.as_bytes())
+            .expect("Failed to write to stdin");
+    });
```

Comment:
I originally suggested doing it this way, but this turns out to be a bit unfortunate, having to spawn a process just to write a constant file. I thought a bit about this and I think it could be nicer by letting Nix write the file to the `/nix/store` and passing its path to the program, which we can do by changing the `default.nix` like this:

```diff
diff --git a/pkgs/test/nixpkgs-check-by-name/default.nix b/pkgs/test/nixpkgs-check-by-name/default.nix
index 0d4e2fcc48a7..006f6971c337 100644
--- a/pkgs/test/nixpkgs-check-by-name/default.nix
+++ b/pkgs/test/nixpkgs-check-by-name/default.nix
@@ -4,9 +4,13 @@
   nix,
   rustfmt,
   clippy,
+  makeWrapper,
   mkShell,
 }:
 let
+
+  runtimeExprPath = "${./src/eval.nix}";
+
   package =
     rustPlatform.buildRustPackage {
       name = "nixpkgs-check-by-name";
@@ -16,7 +20,9 @@ let
         nix
         rustfmt
         clippy
+        makeWrapper
       ];
+      env.NIX_CHECK_BY_NAME_EXPR_PATH = runtimeExprPath;
       # Needed to make Nix evaluation work inside the nix build
       preCheck = ''
         export TEST_ROOT=$(pwd)/test-tmp
@@ -34,7 +40,12 @@ let
         cargo fmt --check
         cargo clippy -- -D warnings
       '';
+      postInstall = ''
+        wrapProgram $out/bin/nixpkgs-check-by-name \
+          --set NIX_CHECK_BY_NAME_EXPR_PATH "$NIX_CHECK_BY_NAME_EXPR_PATH"
+      '';
       passthru.shell = mkShell {
+        env.NIX_CHECK_BY_NAME_EXPR_PATH = runtimeExprPath;
         inputsFrom = [ package ];
       };
     };
```

This uses:
- `.env` to set `NIX_CHECK_BY_NAME_EXPR_PATH` for the build and the shell
- `wrapProgram --set` to propagate the environment variable to the resulting binary

Having this environment variable available in the program, you can pass its value as the argument instead of `-`.

Not only does this allow getting rid of the `spawn`, but it also makes it much easier to reproduce the command, since people can can the exact same expression via the store path in the argument :D

Reply:
Yea sounds good, I'll make the changes. One thing, can you describe what `makeWrapper` does? I tried looking through it but am not very used to jumping around nix code right now (not very good with nix also :( ).   
Just need to know the intention of that specific part, so that I can figure out what the code is doing easily. The whole `wrapProgram` thing is slightly confusing, on first sight, I probably would have simply set the env and hoped the executable would be called with it...

Reply:
Ah yes, there are [docs for `wrapProgram`](https://nixos.org/manual/nixpkgs/unstable/#fun-makeWrapper), which just calls `makeWrapper` (the shell function, not the Nix expression) underneath, for which there are also [docs](https://nixos.org/manual/nixpkgs/unstable/#fun-makeWrapper). Unfortunately it's not very well documented. The source is better in this case. While the docs mention the source file, I can also recommend using a command like

```
❯ nix-shell -p ripgrep

[nix-shell:~/src/nixpkgs]$ rg 'makeWrapper\(\)'
```

to locate shell commands in Nixpkgs. In this case it points to https://github.com/NixOS/nixpkgs/blob/5d6c73222280b7f69109dfa5c9572f4b74c3453c/pkgs/build-support/setup-hooks/make-wrapper.sh#L10-L35

In short: `wrapProgram foo <args>` replaces the binary `foo` with a shell wrapper that calls the original `foo` underneath, but with some changes according to `<args>`. In this case, I used `--env` to make it have an environment variable set when it runs.

We need this because the build runs in a completely isolated environment, which is shut down after the build is done. After the build, the binary is by default run with whatever environment variables the callers process has, which could be anything, even completely empty!

So if we want the program to also have some environment variable when it runs, it needs to be explicitly propagated from the build.

---

Admittedly this is still not ideal, because it should be possible to embed the Nix store path of the expression into the Rust binary itself, such that it doesn't need to read via an environment variable. This could be done with something like [`include_str!("expr-path")`](https://doc.rust-lang.org/std/macro.include_str.html), where you'd do something like this in the `default.nix` file:

```nix
preConfigure = ''
  echo -n "${runtimeExprPath}" > src/expr-path
'';
```

But this is also not that great because it makes it harder for the development shell to work correctly.

Reply:
Done, thanks for the help. One last thing, can you help me understand `passthru.shell`? I saw the definition for `passthru` in nix manual, the definition itself is trivial and does not really give an intuition on what is going around with that.   
I tested it by removing the environment variable from the expression and it does not come up in `nix-shell`. How the nix-shell is interacting with passthru is not very clear to me from the definition [here](https://nixos.org/manual/nixos/stable/#test-opt-passthru) alone.

Reply:
@infinisil I've added the fix, can you review it please?